Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+
+    /*== Conditional is part of the standard library since C++14, keep it here for C++11 compatibility. ==*/
+    template<bool B, class T, class F> struct conditional { typedef T type; };
+    template<class T, class F> struct conditional<false, T, F> { typedef F type; };
+
+  protected:
+    Nd_type* nd; /*!< \brief The accessed NdFlattener. */
+    const typename Nd_type::Index offset; /*!< \brief Index in the currently accessed layer. */
+    //const typename Nd_type::Index size; /*!< \brief Exclusive upper bound for the next index. */
+    IndexAccumulator_Base(Nd_type* nd, typename Nd_type::Index offset):
+      nd(nd), offset(offset) {}
+
+  };
+
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator : public IndexAccumulator_Base<_N,_Nd_type>{
+  public:
+    static constexpr size_t N = _N;
+    using Nd_type = _Nd_type;
+    typedef IndexAccumulator_Base<_N,_Nd_type> Base;
+
+    template<class ...ARGS> IndexAccumulator(ARGS... args): Base(args...) {}
+
+    /*! The Base of NdFlattener<K> is NdFlattener<K-1>, but do also preserve constness.
+     */
+    using Nd_Base_type = typename Base::template conditional<
+      std::is_const<Nd_type>::value,
+      const typename Nd_type::Base,
+      typename Nd_type::Base
+    >::type;
+    /*! Return type of operator[]. */
+    using LookupType = IndexAccumulator<N-1, Nd_Base_type>;
+
+    /*! \brief Read one more index.
+     * \param[in] i - Index.
+     */
+    LookupType operator[] (typename Nd_type::Index i) {
+      //const typename Nd_type::Index new_offset = this->nd->GetIndices()[this->offset+i];
+      //const typename Nd_type::Index new_size = this->nd->GetIndices()[this->offset+i+1] - new_offset;
+      return LookupType(static_cast<Nd_Base_type*>(this->nd),this->nd->GetIndices()[this->offset+i]);
+    }
+  };
+
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator_Checked : public IndexAccumulator<_N,_Nd_type>{
+  public:
+    static constexpr size_t N = _N;
+    using Nd_type = _Nd_type;
+    typedef IndexAccumulator<_N,_Nd_type> Base;
+
+  protected:
+    /*! Exclusive upper bound for the next index. */
+    const typename Nd_type::Index _size;
+
+  public:
+    template<class ...ARGS>
+    IndexAccumulator_Checked(typename Nd_type::Index size, ARGS... args): Base(args...), _size(size) {}
+
+    /*! Return type of operator[]. */
+    using LookupType = IndexAccumulator_Checked<N-1, typename Base::Nd_Base_type>;
+
+    /*! \brief Read one more index, checking whether it is in the range dictated by the NdFlattener and
+     * previous indices.
+     * \param[in] i - Index.
+     */
+    LookupType operator[] (typename Nd_type::Index i) {
+      if(i>=_size){
+        SU2_MPI::Error("NdFlattener: Index out of range.", CURRENT_FUNCTION);
+      }
+      const typename Nd_type::Index new_offset = this->nd->GetIndices()[this->offset+i];
+      const typename Nd_type::Index new_size = this->nd->GetIndices()[this->offset+i+1] - new_offset;
```

Comment:
auto all you can (we don't want people saying templates are hard to read)

Reply:
Now with the `typename Nd_type::` and `this->` gone, it's readable and I left it.